Fix kubectl get -f <file> -o <nondefault printer> so it prints all items in the file

ncdc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**: Fix kubectl get -f <file> -o <nondefault printer> so it prints all the objects in the file, instead of just the first one. Also add a test for this feature.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #38907 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

cc @AdoHe @deads2k @liggitt @fabianofranz @kubernetes/kubectl @kubernetes/sig-cli-misc

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39038)
<!-- Reviewable:end -->

ncdc commented:
Setting priority to P1 to match issue

liggitt commented on pkg/kubectl/cmd/get.go:
```diff
@@ -324,9 +324,7 @@ func RunGet(f cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args [
-		if singular {
-			obj = infos[0].Object
-		} else {
+		if !singular || len(infos) > 1 {
```
I'm more concerned why singular is true in this case

ncdc replied:
https://github.com/kubernetes/kubernetes/blob/3484c6f67637792a3a91a5c7ff04cb5b28e56c23/pkg/kubectl/resource/builder.go#L716:

```go
singular := !b.dir && !b.stream && len(b.paths) == 1
```

smarterclayton commented:
Singular means "give me a single item instead of a list if you can".  "-f
<list>" falls under "cannot", while "-f <item>" falls under "can".

On Tue, Dec 20, 2016 at 2:53 PM, Kubernetes Submit Queue <
notifications@github.com> wrote:

> Labelling this PR as size/S
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39038#issuecomment-268341644>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_px3oEqraGphsbCbuN7YqIjr2iOxpks5rKDJSgaJpZM4LSN3t>
> .
>

jbeda commented:
This is very confusing.  The definition of Singular isn't defined.  It is really "SingleItemImplied" where the way things were specified that the user might be thinking of a single object.  If, instead, there are multiple returned users should deal with it.

I'd update the documentation and/or rename `IntoSingular` to make this clearer.

Based on reading, it looks like the change is good but confusing.  Not sure if that is what @smarterclayton was suggesting or not.

smarterclayton commented:
SingleItemImplied is accurate.  The user asked for one thing, so should get
an object that corresponds to what they asked for (if you ask for a pod,
you get a pod).  The flipside is if you ask for anything that looks like a
list (kubectl get pods, -f <directory>, -f A -f B), you should *always* get
a list.

If you ask for a single item that happens to be a list, you should get a
list.

On Tue, Dec 20, 2016 at 5:27 PM, Joe Beda <notifications@github.com> wrote:

> This is very confusing. The definition of Singular isn't defined. It is
> really "SingleItemImplied" where the way things were specified that the
> user might be thinking of a single object. If, instead, there are multiple
> returned users should deal with it.
>
> I'd update the documentation and/or rename IntoSingular to make this
> clearer.
>
> Based on reading, it looks like the change is good but confusing. Not sure
> if that is what @smarterclayton <https://github.com/smarterclayton> was
> suggesting or not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39038#issuecomment-268377492>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p0rkd4ayZlN8tHy-_oqOtWfF1txKks5rKFZGgaJpZM4LSN3t>
> .
>

AdoHe commented:
We did get a list, and we will flatten this list, handle the item individually, so if `singular` is true, we just print the first object.

ncdc commented:
Are we saying we need to rename IntoSingular as part of this PR?

jbeda commented:
I *think* the behavior we want is this:
1. If the user gives us something that ~looks like a single object and actually *is* a single object then print it as a single object.
2. If a user give us something that ~looks like a single object but is actually a list, then print a list.
3. If a user gives us something that looks like a list (recursive directory, etc) then *always* return a list.

The confusion in the code is that `singular` is meant to talk about what the input seems to imply about what the user wants while it is being interpreted in the code as to the number of items returned.

As this is probably not the last time folks will be confused here it might be better to clear up the interface.

AdoHe commented:
```
The confusion in the code is that singular is meant to talk about what the input seems to imply about what the user wants while it is being interpreted in the code as to the number of items returned.
```
Yes, this is just the truth.

AdoHe commented:
I think I will carry on this one.

ncdc commented:
@AdoHe are you still planning on taking over? If not, I can come back to it.

smarterclayton commented:
I'm fine with a rename.

ncdc commented:
Ok, cool, I'm picking this back up and will do the rename. FYI @AdoHe

ncdc commented:
@fabianofranz @smarterclayton @AdoHe updated, ptal

liggitt commented on pkg/generated/bindata.go:
```diff
@@ -75,7 +75,7 @@ func (fi bindataFileInfo) Sys() interface{} {
 }
 
-var _translationsKubectlDefaultLc_messagesK8sMo = []byte("\x1f\x8b\x08\x00\x00\x09\x6e\x88\x00\xff\xbc\x53\xc1\x6e\xd3\x40\x10\x75\x21\x5c\x7c\xe4\xcc\x61\x90\x40\x2a\x42\x13\xec\xa4\x54\x91\x43\x10\x4a\x68\x51\xa4\x44\x44\x25\x45\x1c\xd9\x78\xa7\xce\xa2\xf5\xae\xb5\xbb\x6e\xda\x23\x1f\xc1\x2f\x70\x81\xbf\xe0\x17\x38\xf3\x2d\x8c\x63\x48\xe1\xd6\x4b\x3b\xf2\xf3\xcc\xac\xdf\xbe\xd9\xd9\x5d\xff\xba\ ...
+var _translationsKubectlDefaultLc_messagesK8sMo = []byte("\x1f\x8b\x08\x00\x00\x09\x6e\x88\x00\xff\xbc\x93\xcf\x6e\xd3\x4e\x10\xc7\x27\xbf\xf6\x77\xf1\x91\x33\x87\x41\x02\xa9\x08\x8d\x59\xbb\xa5\xaa\x5c\x82\x50\x4b\x8b\x22\x25\x6a\x54\x5c\xd4\x23\x1b\xef\xd4\x59\xb4\xd9\xb5\x76\xd7\x4d\x7b\xe4\x21\x78\x05\x2e\xf0\x16\xbc\x02\x67\x9e\x05\x39\x86\x14\x6e\xbd\xb4\x7b\xb0\xfc\x1d\x7d\xfc\x9d\x3f\x9e\ ...
```
revert this file?

ncdc replied:
I dunno - it got updated when I ran make.

ncdc replied:
Ok, undone

Commit:
Fix kubectl get -f <file> -o <nondefault printer>

Fix kubectl get -f <file> -o <nondefault printer> so it prints all the
objects in the file, instead of just the first one. Also add a test for
this feature.

liggitt commented:
seems reasonable to me

ncdc commented:
Rebased, PTAL!

ncdc commented:
Lowering priority so #39475 can go in first

ncdc commented:
Re-adding release-note label because this fixes a regression in 1.5
